### PR TITLE
feat: add precision to IntervalDay and new IntervalCompound type

### LINKED ISSUE #664: Support for an interval type covering both months and (sub)seconds
Substrait currently defines two interval types, [IntervalYearToMonth](https://github.com/substrait-io/substrait/blob/f9e5f9c515d4b8be079bc7d9dfcd89a6fa5e6c7e/proto/substrait/algebra.proto#L876) which contains a number of years and months, and [IntervalDayToSecond](https://github.com/substrait-io/substrait/blob/f9e5f9c515d4b8be079bc7d9dfcd89a6fa5e6c7e/proto/substrait/algebra.proto#L881).

Neither of these types is capable of expressing full intervals that combine years, months, days, minutes, seconds, etc. (For YearToMonth that's probably clear; for DayToSecond one could in theory use large numbers for days but given the number of days in a month/year is not unique, there is a semantic difference between an interval of "40 DAYS 1 MINUTE" and "1 MONTH 10 DAYS 1 MINUTE"

- Spark has an [CalendarIntervalType](https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/types/CalendarIntervalType.html) that it uses for this purpose, 
- Arrow (and by proxy DataFusion) has [IntervalMonthDayNano](https://docs.rs/arrow/latest/arrow/datatypes/struct.IntervalMonthDayNano.html)
- [Postgres](https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-interval/) says `Internally, PostgreSQL stores interval values as months, days, and seconds. The months and days values are integers while the seconds field can have fractions`
- [Duckdb](https://duckdb.org/docs/sql/functions/interval.html) doesn't go into details of implementation but the docs seem like their interval supports a combination of all magnitudes, and they have the same [issue when converting substrait](https://github.com/duckdb/substrait/blob/ccf848f6ab4a4673dc93413080fc640796c09b86/src/to_substrait.cpp#L186) 

Would you be open to adding an third interval type that combines three numbers: months (i32), days (i32), (micro)seconds (i32 for micro or decimal for seconds)?

### PR DESCRIPTION
**Pull request by Blizzara**
* Update IntervalDay to support multiple subsecond precisions (0-9). This is done in a way that should be effectively compatible with old systems. Old plans should be able to be continue to be consumed treating the old values as microsecond precision.
* Add new IntervalCompound type which is a combination of IntervalMonth and IntervalDay

BREAKING CHANGE: The encoding of IntervalDay literals has changed in a strictly backwards incompatible way. However, the logical meaning across encoding is maintained using a oneof. Moving a field into a oneof makes unset/set to zero unclear with older messages but the fields are defined such that the logical meaning of the two is indistinct. If neither microseconds nor precision is set, the value can be considered a precision 6 value. If you aren't using IntervalDay type, you will not need to make any changes.

BREAKING CHANGE: TypeExpression and Parametertized type protobufs (used to serialize output derivation) are updated to match the now compound nature of IntervalDay. If you use protobuf to serialize output derivation that refer to IntervalDay type, you will need to rework that logic.

fixes #664

### CODE REVIEW

File: proto/substrait/parameterized_types.proto
Comment (Blizzara) [inline proto/substrait/parameterized_types.proto:30]: I aligned with the existing names, but personally I think I'd prefer `IntervalMonthDayNano` or similar to be more descriptive - thoughts?

File: proto/substrait/algebra.proto
Comment (Blizzara) [inline proto/substrait/algebra.proto:813]: I aligned with existing names, but the `To` feels weird to me - I guess it indicates the literal contains data for precisions from X to Y, but I feel these would be clearer without it. Can't rename the existing ones w/o breaking things, but should this be `IntervalMonthDayNano` ?

File: proto/substrait/algebra.proto
Comment (Blizzara) [inline proto/substrait/algebra.proto:891]: I opted for nanoseconds here for simplicity, but alternative is to use a variable precision instead like PrecisionTimestamp(Tz) - preferences?

File: proto/substrait/parameterized_types.proto
Comment (vbarua) [reply proto/substrait/parameterized_types.proto:30]: `IntervalMonthDayNano` or some variant seems reasonable (and better) here.

I don't think it makes sense to align with the existing names. If anything I want
* IntervalYear -> IntervalYearToMonth
* IntervalDay -> IntervalDayToSecond

to better align with the other names in the file, but that's not a backwards compatible change.

**Review by jacques-n — CHANGES_REQUESTED** [review]
I've posted my main comments on #664 . One more cosmetic comment here.

File: site/docs/types/type_classes.md
Comment (Blizzara) [reply site/docs/types/type_classes.md]: Done! (otherwise I'll wait for consensus on #664 before changing the impl here)

**Comment by jacques-n** [comment]
Per other comment, I think the discussion is complete in #664. Would be great to get PR update.

**Review by jacques-n — APPROVED** [review]

**Comment by jacques-n** [comment]
I've updated the PR here. @westonpace @EpsilonPrime @vbarua , can you all take a look? 

This is technically a breaking change although the way I did it, it isn't logically one. Nonetheless, I think we should have sufficient +1s before merging.

File: proto/substrait/algebra.proto
Comment (EpsilonPrime) [inline proto/substrait/algebra.proto:886]: I don't think we need the comment (a reader of protobufs should know how oneof works).

File: proto/substrait/algebra.proto
Comment (EpsilonPrime) [inline proto/substrait/algebra.proto:893]: Squash the double space into a single one here?

File: proto/substrait/type.proto
Comment (EpsilonPrime) [inline proto/substrait/type.proto:138]: The part about being non-backwards belongs in the PR and probably not the future implementation.  I suggest removing it.

File: proto/substrait/type.proto
Comment (EpsilonPrime) [inline proto/substrait/type.proto:142]: Does this also contain years?

File: proto/substrait/algebra.proto
Comment (EpsilonPrime) [inline proto/substrait/algebra.proto:897]: Essentially this is struct<iyear, iday>.  To specify it in text one would need to use the nested struct syntax.  Would this be better flattened with the fields from both types included at the top level?

File: proto/substrait/algebra.proto
Comment (jacques-n) [reply proto/substrait/algebra.proto:886]: moved things around a little. the pairing of precision with subseconds is the part that is a little unusual with oneof.

File: proto/substrait/algebra.proto
Comment (jacques-n) [reply proto/substrait/algebra.proto:897]: Are you talking about the JSON representation or something else? I wasn't really worried about the json representation. Felt like this better clearly communicated the nature as opposed to field copying. It's a weak preference though.

Comment (EpsilonPrime) [reply proto/substrait/algebra.proto:896]: Both JSON and the Substrait Text format (which was what I was thinking about since it'll require implementation) would be affected.  I prefer struct<iyear, iday> to the compound approach.  I would also classify my preference as weak.

**Review by EpsilonPrime — DISMISSED** [review]

File: proto/substrait/algebra.proto
Comment (jacques-n) [reply proto/substrait/algebra.proto:896]: I would expect the substrait text format might have the fields be flattened. I don't think it is critical for it to follow the same structure as the protobufs as long as translation from one to the other is supported, right?

**Review by jacques-n — APPROVED** [review]

**Review by westonpace — APPROVED** [review]
We could potentially make this a non-breaking change if we documented something like:

```
If the microseconds value is non-zero then ignore the other two fields and assume microsecond precision.
```

However, that burdens consumers with (a slight amount) of technical debt.

From a quick glance it appears that:

 * Arrow is incapable of producing `IntervalDayToSecond`
 * Arrow is consuming `IntervalDayToSecond` incorrectly already (it is completely ignoring the microseconds field)
 * Datafusion can produce `IntervalDayToSecond` with a microseconds value (this change will break anyone using this behavior)
 * Datafusion consumes `IntervalDayToSecond` but, since DF only has millisecond precision, it silently discards some precision. (this change will break users relying on this behavior but at least we have some positive benefit here of turning that silent refusal into a noisy refusal).

Overall, I think I'm on the fence but will lean towards progress.

**Comment by jacques-n** [comment]
Given @westonpace seeing a lack of clarity about the fact that the literal is backwards compatible, I added an additional comment to the IntervalDay literal to better clarify that consumers need to understand that two different variations are supported. @westonpace and @EpsilonPrime, if you're both good, I'll merge this. Thanks

**Review by westonpace — APPROVED** [review]

**Review by jacques-n — APPROVED** [review]

**Comment by jacques-n** [comment]
Just need an approval from @vbarua or @EpsilonPrime and/or @cpcloud and then we can merge.

**Review by EpsilonPrime — APPROVED** [review]
I thought about it and the text format should be fine reaching down into the other two types as an implementation detail.  This approach is good enough.

**Comment by vbarua** [comment]
Taking a look at this today. I want to verify that backwards compatibility of this internally.

File: site/docs/types/type_classes.md
Comment (vbarua) [inline site/docs/types/type_classes.md:46]: > across multiple grains

What is a grain in this context?

File: site/docs/types/type_classes.md
Comment (vbarua) [inline site/docs/types/type_classes.md:45]: Now that this is a parameterised type, do we need to update all the function extensions that refer to it as `interval_day` to use `interval_day<P>` instead?

Comment (jacques-n) [reply site/docs/types/type_classes.md:45]: Good point. I'm inclined to do in a follow-along once this is merged.

**Review by vbarua — APPROVED** [review]
Overall looks good to me!

File: site/docs/types/type_classes.md
Comment (jacques-n) [reply site/docs/types/type_classes.md:46]: year/month vs day/second. there are no consistent reliable translations between each.

**Pull request merged**
